homeassistant/components/qbittorrent/sensor.py

```diff
@@ -98,55 +120,89 @@ def format_speed(speed):
     return round(kb_spd, 2 if kb_spd < 0.1 else 1)
 
 
-class QBittorrentSensor(SensorEntity):
+class QBittorrentUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching data from the QBittorrent API."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        client: Client,
+        scan_interval: timedelta,
+    ) -> None:
+        """Initialize."""
+        self.client = client
+
+        super().__init__(hass, _LOGGER, name=DOMAIN, update_interval=scan_interval)
+
+    def fetch_data(self) -> dict[str, Any]:
+        """Fetch data from qBittorrent."""
+        try:
+            data = self.client.sync_main_data()
+            return data
+        except RequestException as exception:
+            _LOGGER.error("Connection lost")
+            raise UpdateFailed() from exception
+
+    async def _async_update_data(self) -> dict[str, Any]:
+        """Async update wrapper."""
+        return await self.hass.async_add_executor_job(self.fetch_data)
+
+
+class QBittorrentSensor(CoordinatorEntity[QBittorrentUpdateCoordinator], SensorEntity):
     """Representation of an qBittorrent sensor."""
 
     def __init__(
         self,
         description: SensorEntityDescription,
-        qbittorrent_client,
+        coordinator: QBittorrentUpdateCoordinator,
         client_name,
-        exception,
-    ):
+    ) -> None:
         """Initialize the qBittorrent sensor."""
+        super().__init__(coordinator)
         self.entity_description = description
-        self.client = qbittorrent_client
-        self._exception = exception
 
         self._attr_name = f"{client_name} {description.name}"
-        self._attr_available = False
-
-    def update(self) -> None:
-        """Get the latest data from qBittorrent and updates the state."""
-        try:
-            data = self.client.sync_main_data()
-            self._attr_available = True
-        except RequestException:
-            _LOGGER.error("Connection lost")
-            self._attr_available = False
-            return
-        except self._exception:
-            _LOGGER.error("Invalid authentication")
-            return
 
+    @property
+    def native_value(self) -> StateType:
+        """Return the state."""
+        data = self.coordinator.data
         if data is None:
-            return
+            return None
 
         download = data["server_state"]["dl_info_speed"]
         upload = data["server_state"]["up_info_speed"]
 
```

Comment:
I think it would be nice if you could also add events for torrent start/pause/stop/completed…

https://developers.home-assistant.io/docs/integration_events

Reply:
That might indeed be a usefull future adition, but I believe this to be out of scope of the current PR.

As the proposed changes stand, a user could also create an automation that is triggered when the sensor attributes change and use that to check for each of the torrent start/pause/stop/completed state changes.